sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteTables.java

```diff
@@ -70,82 +85,133 @@
       PCollectionView<String> jobIdToken,
       WriteDisposition writeDisposition,
       CreateDisposition createDisposition,
+      List<PCollectionView<?>> sideInputs,
       DynamicDestinations<?, DestinationT> dynamicDestinations) {
     this.singlePartition = singlePartition;
     this.bqServices = bqServices;
     this.jobIdToken = jobIdToken;
     this.firstPaneWriteDisposition = writeDisposition;
     this.firstPaneCreateDisposition = createDisposition;
+    this.sideInputs = sideInputs;
     this.dynamicDestinations = dynamicDestinations;
+    this.mainOutputTag = new TupleTag<>("WriteTablesMainOutput");
+    this.temporaryFilesTag = new TupleTag<>("TemporaryFiles");
   }
 
-  @StartBundle
-  public void startBundle(StartBundleContext c) {
-    // Clear the map on each bundle so we can notice side-input updates.
-    // (alternative is to use a cache with a TTL).
-    jsonSchemas.clear();
-  }
+  @Override
+  public PCollection<KV<TableDestination, String>> expand(
+      PCollection<KV<ShardedKey<DestinationT>, List<String>>> input) {
+    PCollectionTuple writeTablesOutputs = input.apply(ParDo.of(new WriteTablesDoFn())
+        .withSideInputs(sideInputs)
+        .withOutputTags(mainOutputTag, TupleTagList.of(temporaryFilesTag)));
 
-  @ProcessElement
-  public void processElement(ProcessContext c) throws Exception {
-    dynamicDestinations.setSideInputAccessorFromProcessContext(c);
-    DestinationT destination = c.element().getKey().getKey();
-    TableSchema tableSchema;
-    String jsonSchema = jsonSchemas.get(destination);
-    if (jsonSchema != null) {
-      tableSchema = BigQueryHelpers.fromJsonString(jsonSchema, TableSchema.class);
-    } else {
-      tableSchema = dynamicDestinations.getSchema(destination);
-      if (tableSchema != null) {
-        jsonSchemas.put(destination, BigQueryHelpers.toJsonString(tableSchema));
-      }
-    }
-
-    TableDestination tableDestination = dynamicDestinations.getTable(destination);
-    TableReference tableReference = tableDestination.getTableReference();
-    if (Strings.isNullOrEmpty(tableReference.getProjectId())) {
-      tableReference.setProjectId(
-          c.getPipelineOptions().as(BigQueryOptions.class).getProject());
-      tableDestination = new TableDestination(
-          tableReference, tableDestination.getTableDescription());
-    }
+    // Garbage collect temporary files.
+    // We mustn't start garbage collecting files until we are assured that the WriteTablesDoFn has
+    // succeeded in loading those files and won't be retried. Otherwise, we might fail part of the
+    // way through deleting temporary files, and retry WriteTablesDoFn. This will then fail due
+    // to missing files, causing either the entire workflow to fail or get stuck (depending on how
+    // the runner handles persistent failures).
+    writeTablesOutputs
+        .get(temporaryFilesTag)
+        .setCoder(StringUtf8Coder.of())
+        .apply(WithKeys.<Void, String>of((Void) null))
```

Comment:
<!--thread_id:cc_140850079_t; commit:ea7fcc4c7239456bb792eb7f91c56453c340344e; resolved:0-->
<!--section:context-quote-->
> **jkff** wrote:
> Explain why we group with a void key (this is because GCTemporaryFiles takes an Iterable, right? Is there any reason why we want it to continue taking an Iterable, as opposed to taking single elements and batching the deletions within a bundle?)
> 
> And also explain why we window into global windows and trigger "as fast as possible" - is this because we want file deletion to happen as fast as possible? But isn't the input already globally windowed? Perhaps you need Window.configure().triggering... to emphasize that you're not changing the windowing strategy?

<!--section:body-->
It's because we need a GBK, so we need _some_ key to give.

Also Window.configure().triggering - is that allowed? It's not really documented anywhere.